### PR TITLE
add class weights

### PR DESCRIPTION
**Pull request by kadarakos**
I closed the old class-weights PR #614, because it was way outdated compared to the completely refactored `v9` cross-entropy. This new PR adds the same functionality and a bit more flexibility.

All cross-entropy loss functions take a `class_weights` now, which is is used to compute `sample_weights` for each sample i.e.: the gradient and loss for each sample is multiplied by the `class_weight` corresponding to the target class. 

The `CategoricalCrossentropy` can take `class_weights` as a `Dict[int, float]` or a `Floats1d` and if its a `Dict` it checks the types and converts to `Floats1d`. The `SparseCategoricalCrossentropy` in addition also allows for `Dict[str, float]` in which case it requires `names` to be provided and checks against them. 

It also includes a convenience function to compute `class_weights` based on a commonly used formula e.g. also used in [`sklearn`](https://scikit-learn.org/stable/modules/generated/sklearn.utils.class_weight.compute_class_weight.html). I also went for a dynamic approach here where it takes `label_data: Dict[Hashable, int]`, where the keys are the labels and values are the label counts. It returns `Dict[Hashable, float]` the weight for each label. 

Its not expected for `class_weights` to always be useful, but can be helpful sometimes when there is label imbalance and the relevant performance metric is balance sensitive. I think it makes `spaCy` a more complete library if we can engage this functionality.

### CODE REVIEW

**Comment by kadarakos** [comment]
Does not include complete tests yet.

**Comment by kadarakos** [comment]
Has the tests now!

File: thinc/loss.py [hunk 1]
```diff
@@ -67,10 +71,30 @@ def _validate_input(self, guesses: FloatsOrRaggedT, target: Floats2d) -> None:
                 "with truth values outside the [0,1] interval."
             )
 
+    def _get_sample_weights(self, target: Floats2d) -> Floats1d:
```
[the remark this answers is not in the document]
Comment (kadarakos) [reply thinc/loss.py:74]: I think it makes sense but not sure how well it composes, because I think it would require a fine-grained type system to actually make it fool proof. What I'm thinking is that let's say we implement a `ClassWeighter` that takes a loss as input, but then this loss has to have type `CategoricalLoss`, which could be the cross-entropy or KL or some margin-based etc. Not only that, but it also doesn't compose with `Sequential` type losses only with the internal loss of the `Sequential` loss. 

So we I can go ahead and implement this `ClassWeighter` outside of the cross-entropy losses, but it will just factor out the code for validating and converting the `class_weights` + adjusting the gradients, but I don't think we can make it into this nice general thing.

File: thinc/loss.py [hunk 2]
```diff
@@ -41,6 +42,9 @@ def get_loss(self, guesses: GuessT, truths: TruthT) -> LossT:
 
 class CategoricalCrossentropyBase(Loss):
     normalize: bool
+    # Not exactly sure what type would make sense because the two cross-entropies
```
[the remark this answers is not in the document]
Comment (kadarakos) [reply thinc/loss.py:45]: Not necessarily, because for the sparse version you should be able to provide `Dict[str, float]` and `Dict[int, float]`.

File: thinc/loss.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (danieldk) [reply thinc/loss.py:74]: > What would work better maybe is to use a CrossEntropyWeighter for now that can only take a loss that is isinstance of CategoricalCrossentropyBase and then do the work?

👍 

Or perhaps even just `CategoricalCrossentropy` and a separate weighter for the sequence case. They could share code.

[the remark this answers is not in the document]
Comment (kadarakos) [reply thinc/loss.py:74]: Oh wait, a problem with the wrapper class is that it will have the wrong type right? I can wrap the `CategoricalCrossentropy` in `CrossEntropyWeighter`, but then `CrossEntropyWeighter` has to be also of type `CategoricalCrossEntropyBase` right?

[the remark this answers is not in the document]
Comment (kadarakos) [reply thinc/loss.py:74]: I've been looking into it today and it doesn't seem to compose too well without refactoring the cross-entropy losses again. The weighting comes in right at the gradient calculation in `_get_grad`

```python
differrence = _to_array(guesses) - target
do weighting here
do masking
do normalization
```
Also in `_get_loss` the sample weights need to be injected before summing the loss terms and normalization. So I think I would need to re-write the API yet again to compose well right?

**Comment by svlandeg** [comment]
@kadarakos : can you resolve the merge conflicts?

**Comment by kadarakos** [comment]
> @kadarakos : can you resolve the merge conflicts?

:+1:

File: thinc/loss.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (danieldk) [reply thinc/loss.py:45]: @kadarakos did you see this comment?

**Comment by kadarakos** [comment]
Closing the PR due to priorities changing, PyTorch will take over `thinc` in `spaCy` and adding more features from the `thinc` side is likely not going to pay off at this stage.
